libraries/AC_AttitudeControl/AC_WeatherVane.cpp

```diff
@@ -0,0 +1,235 @@
+            output = (fabsf(roll_cdeg) + MAX(pitch_cdeg,0.0)) * _gain;
+
+            // Yaw in the direction of the lowest 'wing'
+            if (roll_cdeg < 0) {
+                output *= -1.0;
+            }
+            break;
+
+        case Direction::SIDE_IN:
+            output = pitch_cdeg * _gain;
+
+            if (roll_cdeg < 0) {
+                output *= -1.0;
+            }
+            break;
+
+    }
+
+    // Don't activate weather vane if less than deadzone angle
+    float angle = get_direction() == Direction::SIDE_IN ? pitch_cdeg : roll_cdeg;
+    if (fabsf(angle) < _min_dz_ang_deg*100.0 && !(pitch_cdeg > _min_dz_ang_deg*100.0 && get_direction() == Direction::NOSE_IN)) {
+        output = 0.0;
+    }
+
+    // Force the controller to relax.  This can be called when landing.
+    if (should_relax) {
+        output = 0.0;
+        // Always reset should relax. Maintain relaxed condition by persistant calls to relax()
+        should_relax = false;
+    }
+
+    // Slew output
+    last_output = 0.98 * last_output + 0.02 * output;
+
+    return last_output;
+}
+
+// Called on an interupt to reset the weathervane controller
+void AC_WeatherVane::reset(void)
+{
+    last_output = 0;
+    active_msg_sent = false;
+    should_relax = false;
+    first_activate_ms = 0;
+}
+
+bool AC_WeatherVane::should_weathervane(const int16_t pilot_yaw, const int16_t roll_cdeg, const int16_t pitch_cdeg)
+{
+    // Check enabled
+    if (get_direction() == Direction::OFF){
+        reset();
+        return false;
+    }
+
+    const uint32_t now = AP_HAL::millis();
+
+    // Don't fight pilot inputs
+    if (pilot_yaw != 0) {
+        last_pilot_input_ms = now;
+        reset();
+        return false;
+    }
+
+    // Only allow weather vaning if no input from pilot in last 2 seconds
+    const uint32_t buffer_time_ms = 2000;
+    if (now - last_pilot_input_ms < buffer_time_ms) {
+        reset();
+        return false;
+    }
+
+    // Check if we are above the minimum height to weather vane
+    if (below_min_height()) {
+        reset();
+        return false;
+    }
+
+    // Check if we meet the horizontal velocity thresholds to allow weathervaning
+    if ((_max_vel_xy.get() > 0) && (_inav.get_speed_xy()*0.01) > _max_vel_xy.get()) {
+        reset();
+        return false;
+    }
+
+    // Check if we meet the vertical velocity thresholds to allow weathervaning
+    if ((_max_vel_z.get() > 0) && fabsf(_inav.get_velocity_z())*0.01 > _max_vel_z.get()) {
+        reset();
+        return false;
+    }
+
+    /*
+      Use a 2 second buffer to ensure weathervaning occurs once the vehicle has
+      clearly achieved an acceptable condition.
+    */
+    if (first_activate_ms == 0) {
+        first_activate_ms = now;
+    }
+    if (now - first_activate_ms < buffer_time_ms) {
+        return false;
+    }
+
+    // If we got this far then we should allow weathervaning
+    return true;
+}
+
+bool AC_WeatherVane::below_min_height(void)
```

Comment:
this doesn't look like a match to the existing function, instead pass height above ground from Plane::relative_ground_altitude()